keyring/backends/Windows.py

```diff
@@ -88,79 +301,152 @@ def priority(cls):
         """
         if missing_deps:
-            raise RuntimeError("Requires Windows and pywin32")
+            raise RuntimeError("Requires Windows and cffi")
         return 5
 
     @staticmethod
     def _compound_name(username, service):
-        return f'{username}@{service}'
+        return f"{username}@{service}"
+
+    def _discover_max_password_bytes(self):
+        def test_length(length):
+            try:
+                # NOTE relies on character "a" mapping to a single wchar_t (i.e. 2 bytes)
+                try:
+                    self._set_password("__probe_target__", "username", "a" * length)
+                except Exception as e:
+                    if not (isinstance(e, CredError) and e.winerror == 1783):
+                        log.exception(f"_set_password raised {e=}")
+                    raise
+                self._delete_password("__probe_target__")
+                return True
+            except CredError as e:
+                if e.winerror == 1783 and e.funcname == "CredWrite":
+                    return False
+                else:
+                    log.exception("unexpected CredError")
+                    return False
+            except Exception:
+                log.exception("unexpected Exception")
+
+        # first try to confirm observed max bytes of 2560
+        if test_length(2560) and not test_length(2561):
+            return 2560
+
+        # otherwise, use binary search
+        start, end = 1, MAX_PASSWORD_BYTES // 2
+        while start <= end:
+            mid = (start + end) // 2
+            if test_length(mid):
+                start = mid + 1
+            else:
+                end = mid - 1
+        return end
+
+    def __init__(self, *arg, **kw):
+        super().__init__(*arg, **kw)
+
+        # prevent sharding during self._discover_max_password_bytes
+        self._is_init = True
+        self._max_password_bytes = self._discover_max_password_bytes()
+        self._is_init = False
 
     def get_password(self, service, username):
         # first attempt to get the password under the service name
-        res = self._get_password(service)
-        if not res or res['UserName'] != username:
+        creds = self._get_password(service)
+        if not creds or creds["UserName"] != username:
             # It wasn't found so attempt to get it with the compound name
-            res = self._get_password(self._compound_name(username, service))
-        if not res:
+            creds = self._get_password(self._compound_name(username, service))
+        if not creds:
             return None
-        return res.value
+        password = creds.value
+        return password
 
-    def _get_password(self, target):
+    def _get_password_inner(self, target):
         try:
-            res = win32cred.CredRead(
-                Type=win32cred.CRED_TYPE_GENERIC, TargetName=target
-            )
-        except pywintypes.error as e:
-            e = OldPywinError.wrap(e)
```

Comment:
Glad to see this legacy compatibility go away. Probably this change could be made independently and rolled out in advance to ensure that it doesn't break any use-cases.

Reply:
I don't think I understand your comment.  Are you saying that this change should be made in a separate Pull Request?  I'm not sure how to do that.

Reply:
Yes. That's what I was suggesting. No problem. I committed the change in 96d3995, which I'll merge into this PR.

Reply:
I'd planned to merge the change into this PR, but found I don't have permission to push to your fork. It's a good idea when making a pull request to (a) Create a branch (other than main) in your own fork for the change and (b) grant maintainers of the upstream fork access to that branch. You can pull my merge commit using `git pull https://github.com/jaraco/keyring LiveData-Inc/main`.

Reply:
Clearly I have a lot to learn.  Sorry for my repeated mistakes.  I wish I had read your [blog post](https://blog.jaraco.com/how-to-write-perfect-pull-request/) before I had started!

I will pull your merge commit, and grant you access.  I'm assuming (out of sheer ignorance) that it is too late for me to create a branch etc?

Reply:
No worries, and don't be so hard on yourself ;)

> I'm assuming (out of sheer ignorance) that it is too late for me to create a branch etc?

It's never too late. Except maybe for this specific pull request. But since elsewhere, I've recommended to step back and approach the problem differently, it will be a good opportunity to work with branches. Here's what I'd recommend (I'm going to assume you're using a command-line client, but you can find similar operations in a GUI client):

1. First, you want to save the work you've already done in this PR (in your main branch) by giving it a branch name. Make sure you've committed any changes you want to keep or discard any changes you do not. Then, while still on your main (with the latest changes pulled), just run `git branch feature/windows-shard-pr544`. That will create a new branch with its head pointing to the same commits as your current main. That's not normally how one creates a branch, but it's a good way to create a branch without checking out that branch.
2. Since you're still on `main`, reset it to the state of upstream. If you don't already have an `upstream` remote configured, you should configure one with `git remote add upstream https://github.com/jaraco/keyring`. Then make sure you have the commits from upstream with `git fetch upstream`. Then reset main with `git reset --hard upstream/main`.
3. Now you're ready to create a new branch for new development. Use `git checkout -b feature/windows-shard-take-2` (you can call the branch whatever you want; I just suggest names like I might use).
4. Finally, after you've committed some changes to your branch and you want to create a pull request, you first need to push that branch to your fork. Assuming your fork is the "origin", while on the branch, you do that with `git push --set-upstream origin` (I think; you may also need to pass the name of the branch).

I know this branching stuff can get confusing, so I do recommend grabbing an SCM visualization GUI like Sourcetree (free) or Sublime Merge ($) or PyCharm ($). It helps me visualize the changes that are happening to different branches of development and their state across local and remote repos. If you love the command line, you may also be interested in the [gh tool](https://github.com/cli/cli) as it helps facilitate a lot of github operations (such as creating pull requests) from the command-line.

Hope that helps you get started on your next branch!